pkg/trie/triedb/nibbles/nibbles.go

```diff
@@ -0,0 +1,365 @@
+// is a non expanded byte slice followed by a last padded byte representation.
+// The padded byte is an optional padded value.
+type Prefix struct {
+	Key    []byte
+	Padded *byte
+}
+
+func (p Prefix) JoinedBytes() []byte {
+	if p.Padded != nil {
+		return append(p.Key, *p.Padded)
+	}
+	return p.Key
+}
+
+// Return left portion of [Nibbles], if the slice
+// originates from a full key it will be the Prefix of
+// the node.
+func (n Nibbles) Left() Prefix {
+	split := n.offset / NibblesPerByte
+	ix := uint8(n.offset % NibblesPerByte) //nolint:gosec
+	if ix == 0 {
+		return Prefix{Key: n.data[:split]}
+	}
+	padded := PadLeft(n.data[split])
+	return Prefix{Key: slices.Clone(n.data[:split]), Padded: &padded}
+}
+
+func (n Nibbles) Len() uint {
+	return uint(len(n.data))*NibblesPerByte - n.offset
+}
+
+// Advance the view on the slice by i nibbles.
+func (n *Nibbles) Advance(i uint) {
+	if !(n.Len() >= i) {
+		panic("not enough nibbles to advance")
+	}
+	n.offset += i
+}
+
+// Move back to a previously valid fix offset position.
+func (n Nibbles) Back(i uint) Nibbles {
+	return Nibbles{data: n.data, offset: i}
+}
+
+func (n Nibbles) Equal(them Nibbles) bool {
+	return n.Len() == them.Len() && n.StartsWith(them)
+}
+
+func (n Nibbles) StartsWith(them Nibbles) bool {
+	return n.CommonPrefix(them) == them.Len()
+}
+
+// Calculate the number of common nibble between two left aligned bytes.
+func leftCommon(a uint8, b uint8) uint {
+	if a == b {
+		return 2
+	} else if PadLeft(a) == PadLeft(b) {
+		return 1
+	} else {
+		return 0
+	}
+}
+
+// Count the biggest common depth between two left aligned packed nibble slice.
+func biggestDepth(v1 []uint8, v2 []uint8) uint {
+	upperBound := len(v1)
+	if len(v2) < upperBound {
+		upperBound = len(v2)
+	}
+	for a := 0; a < upperBound; a++ {
+		if v1[a] != v2[a] {
+			return uint(a)*NibblesPerByte + leftCommon(v1[a], v2[a]) //nolint:gosec
+		}
+	}
+	return uint(upperBound) * NibblesPerByte //nolint:gosec
+}
+
+// How many of the same nibbles at the beginning do we match with them?
+func (n Nibbles) CommonPrefix(them Nibbles) uint {
+	selfAlign := n.offset % NibblesPerByte
+	themAlign := them.offset % NibblesPerByte
+	if selfAlign == themAlign {
+		selfStart := n.offset / NibblesPerByte
+		themStart := them.offset / NibblesPerByte
+		var first uint = 0
+		if selfAlign != 0 {
+			if PadRight(n.data[selfStart]) != PadRight(them.data[themStart]) {
+				// warning only for radix 16
+				return 0
+			}
+			selfStart += 1
+			themStart += 1
+			first += 1
+		}
+		return biggestDepth(n.data[selfStart:], them.data[themStart:]) + first
+	} else {
+		s := n.Len()
+		if them.Len() < s {
+			s = them.Len()
+		}
+		var i uint
+		for i < s {
+			if n.At(i) != them.At(i) {
+				break
+			}
+			i++
+		}
+		return i
+	}
+}
+
+// Helper function to create a [NodeKey].
+func (n Nibbles) NodeKey() NodeKey {
+	split := n.offset / NibblesPerByte
+	offset := n.offset % NibblesPerByte
+	return struct {
```

Comment:
Why use an anon struct here instead of `NodeKey`?

Reply:
good catch! I had used anonymous structure before for `NodeKey` but changed it during development.